cmd/k8sagent/unit/manifolds.go

```diff
@@ -0,0 +1,330 @@
+	LeadershipGuarantee time.Duration
+
+	// AgentConfigChanged is set whenever the unit agent's config
+	// is updated.
+	AgentConfigChanged *voyeur.Value
+
+	// ValidateMigration is called by the migrationminion during the
+	// migration process to check that the agent will be ok when
+	// connected to the new target controller.
+	ValidateMigration func(base.APICaller) error
+
+	// PrometheusRegisterer is a prometheus.Registerer that may be used
+	// by workers to register Prometheus metric collectors.
+	PrometheusRegisterer prometheus.Registerer
+
+	// UpdateLoggerConfig is a function that will save the specified
+	// config value as the logging config in the agent.conf file.
+	UpdateLoggerConfig func(string) error
+
+	// PreviousAgentVersion passes through the version the unit
+	// agent was running before the current restart.
+	PreviousAgentVersion version.Number
+
+	// UpgradeStepsLock is passed to the upgrade steps gate to
+	// coordinate workers that shouldn't do anything until the
+	// upgrade-steps worker is done.
+	UpgradeStepsLock gate.Lock
+
+	// PreUpgradeSteps is a function that is used by the upgradesteps
+	// worker to ensure that conditions are OK for an upgrade to
+	// proceed.
+	PreUpgradeSteps func(*state.StatePool, coreagent.Config, bool, bool, bool) error
+
+	// MachineLock is a central source for acquiring the machine lock.
+	// This is used by a number of workers to ensure serialisation of actions
+	// across the machine.
+	MachineLock machinelock.Lock
+
+	// Clock contains the clock that will be made available to manifolds.
+	Clock clock.Clock
+}
+
+// Manifolds returns a set of co-configured manifolds covering the various
+// responsibilities of a k8s agent unit command. It also accepts the logSource
+// argument because we haven't figured out how to thread all the logging bits
+// through a dependency engine yet.
+//
+// Thou Shalt Not Use String Literals In This Function. Or Else.
+func Manifolds(config manifoldsConfig) dependency.Manifolds {
+	return dependency.Manifolds{
+		// The agent manifold references the enclosing agent, and is the
+		// foundation stone on which most other manifolds ultimately depend.
+		agentName: agent.Manifold(config.Agent),
+
+		// The api-config-watcher manifold monitors the API server
+		// addresses in the agent config and bounces when they
+		// change. It's required as part of model migrations.
+		apiConfigWatcherName: apiconfigwatcher.Manifold(apiconfigwatcher.ManifoldConfig{
+			AgentName:          agentName,
+			AgentConfigChanged: config.AgentConfigChanged,
+			Logger:             loggo.GetLogger("juju.worker.apiconfigwatcher"),
+		}),
+
+		apiCallerName: apicaller.Manifold(apicaller.ManifoldConfig{
+			AgentName:            agentName,
+			APIOpen:              api.Open,
+			APIConfigWatcherName: apiConfigWatcherName,
+			NewConnection:        apicaller.OnlyConnect,
+			Logger:               loggo.GetLogger("juju.worker.apicaller"),
+		}),
+
```

Comment:
I think we need to create a worker whose sole job is to set the unit's version. Currently this is done in the worker/upgrader. This unit agent doesn't need it, nor do the units when we merge the agents together, but we will need one. Can I get you to add a comment like this:

```
TODO: make a worker to set agent version and remove from worker/upgrader.
```

And we'll get to it as the merging agents work should also have this.